### ISSUE TITLE
Copying segment iterators is expensive

### ISSUE THREAD

**Issue by mrks** [hyrise/hyrise#1358]
@mrzzzrm had a performance issue here: https://github.com/hyrise/hyrise/pull/1337#issuecomment-445741302

I remember having a similar issue, too. I don't see a good reason why those copies should be expensive. If they are, that either needs to get fixed or we should consider outlawing copies altogether.

**Comment by cmfcmf** [comment]
It'd be interesting to see whether this is still the case after the recent changes in #1515.

**Comment by mrzzzrm** [comment]
We are mainly talking about string copies into SegmentPosition (which is what Iterators get dereferenced into) here.

**Comment by mrzzzrm** [comment]
@mrks, I think you were on this, right? Assigned you for completeness sake. As more people work on Hyrise again, we should probably be diligent with this to avoid work being done twice.

**Comment by mrks** [comment]
> We are mainly talking about string copies into SegmentPosition (which is what Iterators get dereferenced into) here.

I don't see you copying the SegmentPosition in your code though.

**Comment by mrzzzrm** [comment]
Would need to re-check this at some point

**Comment by mrzzzrm** [comment]
`SegmentPosition` holds it's value as an instance, so here string copies will be created.

```c++
template <typename T>
class SegmentPosition : public AbstractSegmentPosition<T> {
 public:
  static constexpr bool Nullable = true;

  SegmentPosition(const T& value, const bool null_value, const ChunkOffset& chunk_offset)
      : _value{value}, _null_value{null_value}, _chunk_offset{chunk_offset} {}

  const T& value() const final { return _value; }
  bool is_null() const final { return _null_value; }
  ChunkOffset chunk_offset() const final { return _chunk_offset; }

 private:
  // The alignment improves the suitability of the iterator for (auto-)vectorization
  alignas(8) const T _value;
  alignas(8) const bool _null_value;
  alignas(8) const ChunkOffset _chunk_offset;
};
```

**Comment by mrks** [comment]
I made multiple attempts at replacing this with `T&`, `T*`, or `string_view`. Out of architectural beauty, performance, and correctness I was only ever able to pick two.

It's good to know where this comes from. Any idea how we could solve this? Otherwise, let's just add a comment and reference to this issue and call it a day.

**Comment by mrks** [comment]
(Not saying that getting rid of `T _value` can't or shouldn't be done. Just proposing to close this before it becomes a zombie issue.)

**Comment by mrks** [comment]
I tried making SegmentPosition non-copyable. The problem is that boost's operator_arrow_dispatch internally still copies the iterator's value unless we make it return a reference. I tried that by copying the value into a temporary stored in the iterator (32e533d1345831c091a5d6edf349683194572b62) but that decreased the performance.

Until we have a good way of referencing the actual base data (which is not even possible for, e.g., LZ4 segments), we will just need to make sure that SegmentPositions are not copied unnecessarily.
